readthedocs/vcs_support/backends/git.py

```diff
@@ -68,6 +91,166 @@ def update(self):
             return self.fetch()
         return self.clone()
 
+    def get_remote_fetch_refspec(self):
+        """
+        Gets a valid remote reference for the identifier.
+
+        See also: The <refspec> section from ``git help fetch``
+
+        This method sits on top of a lot of legacy design.
+        It decides how to treat the incoming ``Version.identifier`` from
+        knowledge of how the caller (the build process) uses build data.
+
+        Version.identifier = a branch name (branches)
+        Version.identifier = commit (tags)
+        Version.identifier = commit (external versions)
+        Version.verbose_name = branch alias, e.g. latest (branches)
+        Version.verbose_name = tag name (tags)
+        Version.verbose_name = PR number (external versions)
+
+        :return: A refspec valid for fetch operation
+        """
+
+        if not self.version_type:
+            log.warning(
+                "Trying to resolve a remote reference without setting version_type is not "
+                "possible",
+                project=self.project,
+            )
+            return None
+
+        # Branches have the branch identifier set by the caller who instantiated the
+        # Git backend.
+        # If version_identifier is empty, then the fetch operation cannot know what to fetch
+        # and will fetch everything, in order to build what might be defined elsewhere
+        # as the "default branch". This can be the case for an initial build started BEFORE
+        # a webhook or sync versions task has concluded what the default branch is.
+        if self.version_type == BRANCH and self.version_identifier:
+            # Here we point directly to the remote branch name and update our local remote
+            # refspec to point here.
+            # The original motivation for putting 'refs/remotes/origin/<branch>' as the local refspec
+            # was an assumption in the .branches property that iterates over everything in
+            # refs/remotes/origin. We might still keep this pattern as it seems to be the most solid
+            # convention for storing a remote:local refspec mapping.
+            return (
+                f"refs/heads/{self.version_identifier}:refs/remotes/origin/"
+                f"{self.version_identifier}"
+            )
+        # Tags
+        if self.version_type == TAG and self.verbose_name:
+            # A "stable" tag is automatically created with Version.machine=True,
+            # denoting that it's not a tag that really exists.
+            # Because we don't know if it originates from the default branch or some
+            # other tagged release, we will fetch everything.
+            if self.version_machine and self.verbose_name == "stable":
+                if self.version_identifier:
+                    return f"{self.version_identifier}:readthedocs/stable"
```

Comment:
Where are we calling the `git checkout readthedocs/stable` command after fetching it? I didn't find it in this PR.

Reply:
We checkout the Git commit hash directly. I can see if we can actually just remove `:readthedocs/stable` - it adds some confusion when we aren't actually using it.

Reply:
Removed, will test :+1: